File: AdminService/src/main/java/com/ecommercial/shopping/adminservice/global/config/DataSourceConfig.java
package com.ecommercial.shopping.adminservice.global.config;

import com.ecommercial.shopping.adminservice.global.database.RoutingDataSource;
import com.zaxxer.hikari.HikariDataSource;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy;

import javax.sql.DataSource;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

@Configuration
@RequiredArgsConstructor
public class DataSourceConfig {
    private final DataSourceProperties dataSourceProperties;

    @Bean
    public DataSource routingDataSource() {
        RoutingDataSource routingDataSource = new RoutingDataSource();

        DataSourceProperties.Write write = dataSourceProperties.getWrite();
        DataSource writeDataSource = createDataSource(write.getUrl());

        Map<Object, Object> dataSourceMap = new HashMap<>();
        dataSourceMap.put(write.getUrl(), writeDataSource);
        System.out.println("0--0=-------");
        System.out.println(write.getUrl());

        List<DataSourceProperties.Read> reads = dataSourceProperties.getReads();

        for (DataSourceProperties.Read read : reads) {
            dataSourceMap.put(read.getName(), createDataSource(read.getUrl()));
        }

        routingDataSource.setDefaultTargetDataSource(writeDataSource);
        routingDataSource.setTargetDataSources(dataSourceMap);
        routingDataSource.afterPropertiesSet();

        return new LazyConnectionDataSourceProxy(routingDataSource);

    }

    private DataSource createDataSource(String url) {
        HikariDataSource hikariDataSource = new HikariDataSource();
        hikariDataSource.setDriverClassName(dataSourceProperties.getDriverClassName());
        hikariDataSource.setUsername(dataSourceProperties.getUsername());
        hikariDataSource.setPassword(dataSourceProperties.getPassword());
        hikariDataSource.setJdbcUrl(url);

        return hikariDataSource;
    }
}

File: UserService/src/main/kotlin/com/ecommercial/shopping/userservice/global/config/DataSourceConfig.kt
package com.ecommercial.shopping.userservice.global.config
import com.zaxxer.hikari.HikariDataSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy

@Configuration
class DataSourceConfig(
    private val dataSourceProperties: DataSourceProperties
) {

    @Bean
    fun routingDataSource(): LazyConnectionDataSourceProxy {
        try {
            println("dsdsds@!@#@")
            val routingDataSource = RoutingDataSource()

            val write = dataSourceProperties.write
            val writeDataSource = createDataSource(write.url)

            val dataSourceMap = HashMap<Any, Any>()
            dataSourceMap[write.url] = writeDataSource

            dataSourceProperties.reads.forEach { read ->
                dataSourceMap[read.name] = createDataSource(read.url)
            }

            routingDataSource.setDefaultTargetDataSource(writeDataSource)
            routingDataSource.setTargetDataSources(dataSourceMap)
            routingDataSource.afterPropertiesSet()

            return LazyConnectionDataSourceProxy(routingDataSource)
        } catch (e: Exception) {
            println(e.message)
            println("@@@@@@@@@@@@@")
            throw e
        }

    }

    private fun createDataSource(url: String): HikariDataSource {
        val driverName : String = dataSourceProperties.driverClassName
        val userName: String = dataSourceProperties.username
        val password: String = dataSourceProperties.password
        return HikariDataSource().apply {
            driverClassName = driverName
            username = userName
            this.password = password
            jdbcUrl = url
        }
    }
}